libcst/metadata/scope_provider.py

```diff
@@ -86,6 +99,84 @@ class BuiltinAssignment(BaseAssignment):
     pass
 
 
+class QualifiedNameSource(Enum):
+    IMPORT = auto()
+    LOCAL = auto()
+
+
+@dataclass(frozen=True)
+class QualifiedName:
+    name: str
+    source: QualifiedNameSource
+
+
+class _QualifiedNameUtil:
+    @staticmethod
+    def get_full_name_for(node: cst.CSTNode) -> Optional[str]:
+        if isinstance(node, cst.Name):
+            return node.value
+        elif isinstance(node, cst.Attribute):
+            return (
+                f"{_QualifiedNameUtil.get_full_name_for(node.value)}.{node.attr.value}"
+            )
+        elif isinstance(node, cst.Call):
+            return _QualifiedNameUtil.get_full_name_for(node.func)
+        elif isinstance(node, cst.Subscript):
+            return _QualifiedNameUtil.get_full_name_for(node.value)
+        elif isinstance(node, cst.SimpleString):
+            # In the case of SimpleString of type hints.
+            return literal_eval(node.value)
```

Comment:
I am wondering if this should be explicitly disallowed. It might be surprising that people accidentally pass a string instead of a variable the string is assigned to, and find out that it crashes during literal_eval. I guess it is fine as long as we are documenting it though. Thoughts?

Reply:
This is a feature supported in the ast scope API. I guess it's used in some lint use cases for typing.
In the use case of typing, the application may want to know the qualified name of a type object or simple string to verify something.
@bgw do you have comments?

Reply:
You could accept the wrapping `libcst.Annotation` instead of a string, but then it's less useful because there are some cases where you can have an annotation outside of a `libcst.Annotation`, e.g.:

```
T = TypeVar("T", bound="List[int]")
```

or even

```
# This a type alias, but it's hard to tell from the CST.
# It might not even be used in this module, but instead exported
# and used as a type alias by other modules.
MyTypeAlias = "List[Union[int, str]]"
```

It's impossible to figure out every place a string might be used as a type annotation without doing a full cross-module type-check first. Even pyre struggles with this.

I think since we don't have a clear idea of what to do here, maybe we should just drop this functionality. It's always easier to add functionality later than it is to remove or fix buggy functionality (without breaking backwards compatibility).

I think it's fair to document that string-wrapped types aren't supported by this API.

Reply:
Dropped the support of SimpleString and add related document/test for now.